File: AcmeHrApp/backend/graph_ql/queries.py
import graphene
from graphene import Field, List
from .all_type_defs.login import LoginType
from .all_type_defs.get_my_information import MyInformationType
from .all_type_defs.get_my_departments import MyDepartmentInfoType
from .all_type_defs.get_my_organisation import MyOrganisationType
from .all_type_defs.get_employee_details import EmployeeInformationType
from .all_type_defs.get_paginated_colleagues import PaginatedColleaguesType
from .all_queries.user_queries.login import login_query
from .all_queries.user_queries.get_my_information import get_my_information_query
from .all_queries.hr_manager_queries.get_my_organisation_info import get_my_organisation_info_query
from .all_queries.managers_queries.get_my_department_info import get_my_department_info_query
from .all_queries.managers_queries.get_employee_details import get_employee_details_query
from .all_queries.user_queries.get_next_group_of_colleagues import get_next_group_of_colleagues_query


class Query(graphene.ObjectType):
    login = Field(LoginType, employee_id=graphene.String(
        required=True), password=graphene.String(required=True))
    get_my_information = Field(MyInformationType, employee_id=graphene.Int(
        required=True))
    get_my_department_info = Field(MyDepartmentInfoType, dept_no=graphene.String(
        required=True))
    get_my_organisation_info = Field(MyOrganisationType)
    get_employee_details = Field(EmployeeInformationType, emp_no=graphene.Int(
        required=True), accessibility=graphene.String(
        required=True))
    get_paginated_colleagues = Field(PaginatedColleaguesType, cursor=graphene.Int(
        required=True))

    def resolve_login(root, info, employee_id, password):
        response = login_query(employee_id, password)
        return response

    def resolve_get_my_information(root, info, employee_id):
        response = get_my_information_query(employee_id)
        return response

    def resolve_get_my_department_info(root, info, dept_no):
        response = get_my_department_info_query(dept_no)
        return response

    def resolve_get_my_organisation_info(root, info,):
        dept_with_managers_list = get_my_organisation_info_query()
        return dept_with_managers_list

    def resolve_get_employee_details(root, info, emp_no, accessibility):
        response = get_employee_details_query(emp_no, accessibility)
        return response

    def resolve_get_paginated_colleagues(root, info, cursor):
        response = get_next_group_of_colleagues_query(cursor)
        return response

File: frontend/src/graphql/queries.js
import { gql } from "@apollo/client";

const LOGIN_QUERY = `
query login($employeeId: String! $password: String!) {
  login(employeeId: $employeeId password: $password) {
  	colleagues(first: 1000){
    	pageInfo {
      	startCursor
      	endCursor
      	hasNextPage
      	hasPreviousPage
      	}
      	edges {
        	cursor
        	node {
          	title {
            	title
            	toDate
          	}
          	birthDate
          	empNo
          	firstName
          	lastName
          	gender
          	hireDate
        	}
      	}
    	}
    user {
      userInfo {
        birthDate
        empNo
        firstName
        lastName
        gender
        hireDate
      }
      deptNo
      accessibility
    }
    totalEmployeeCount
	}
}`;

const LOGIN = gql`
  ${LOGIN_QUERY}
`;

const GET_PAGINATED_COLLEAGUES_QUERY = `
query getPaginatedColleagues($cursor: Int!) {
  getPaginatedColleagues(cursor: $cursor) {
    hasNextPage
    colleagues {
      pageInfo {
      	startCursor
      	endCursor
      	hasNextPage
      	hasPreviousPage
      	}
      	edges {
        	cursor
        	node {
          	title {
            	title
            	toDate
          	}
          	birthDate
          	empNo
          	firstName
          	lastName
          	gender
          	hireDate
        	}
      	}
    }
  }
}
`;

const GET_PAGINATED_COLLEAGUES = gql`
  ${GET_PAGINATED_COLLEAGUES_QUERY}
`;

const GET_EMPLOYEE_DETAILS_QUERY = `
query getEmployeeDetails($empNo: Int!, $accessibility: String!) {
  getEmployeeDetails(empNo: $empNo, accessibility: $accessibility) {
    deptInfo {
      departments {
        deptName
        deptNo
      }
      deptEmp {
        empNo
        fromDate
        toDate
        deptNo
      }  
    }
    salaryInfo {
      empNo
      salary
      fromDate
      toDate
    }
    titleInfo {
      empNo
      title
      fromDate
      toDate
    }
    user {
      userInfo {
        birthDate
        empNo
        firstName
        lastName
        gender
        hireDate
      }
    }
    employeesManager {
      birthDate
      empNo
      firstName
      lastName
      gender
      hireDate
    }
  }
}
`;

const GET_EMPLOYEE_DETAILS = gql`
  ${GET_EMPLOYEE_DETAILS_QUERY}
`;

const GET_MY_INFORMATION_QUERY = `
query getMyInformation($employeeId: Int!) {
  getMyInformation(employeeId: $employeeId) {
    deptInfo {
      departments {
        deptName
        deptNo
      }
      deptEmp {
        empNo
        fromDate
        toDate
        deptNo
      }  
    }
    salaryInfo {
      empNo
      salary
      fromDate
      toDate
    }
    titleInfo {
      empNo
      title
      fromDate
      toDate
    }
  }
}
`;

const GET_MY_INFORMATION = gql`
  ${GET_MY_INFORMATION_QUERY}
`;

const GET_MY_DEPARTMENT_INFO_QUERY = `
query getMyDepartmentInfo($deptNo: String!) {
  getMyDepartmentInfo(deptNo: $deptNo) {
    department {
      deptName
      deptNo
    }
    departmentEmployees {
    	pageInfo {
      	startCursor
      	endCursor
      	hasNextPage
      	hasPreviousPage
      	}
      	edges {
        	cursor
        	node {
          	title {
            	title
            	toDate
          	}
          	birthDate
          	empNo
          	firstName
          	lastName
          	gender
          	hireDate
        	}
      	}
    	}
  }
}
`;

const GET_MY_DEPARTMENT_INFO = gql`
  ${GET_MY_DEPARTMENT_INFO_QUERY}
`;

const GET_MY_ORGANISATION_INFO_QUERY = `
query getMyOrganisationInfo {
  getMyOrganisationInfo {
    deptWithManagersList{
      department {
      deptNo_DeptName
      deptNo
      total
      }
      manager {
        empNo
        empNo_FirstName
        empNo_LastName
      }
    }
    titlesList {
      title
      }
    }
  }
`;

const GET_MY_ORGANISATION_INFO = gql`
  ${GET_MY_ORGANISATION_INFO_QUERY}
`;

export {
  LOGIN_QUERY,
  GET_MY_DEPARTMENT_INFO_QUERY,
  GET_MY_ORGANISATION_INFO_QUERY,
  LOGIN,
  GET_PAGINATED_COLLEAGUES_QUERY,
  GET_EMPLOYEE_DETAILS_QUERY,
  GET_EMPLOYEE_DETAILS,
  GET_PAGINATED_COLLEAGUES,
  GET_MY_INFORMATION_QUERY,
  GET_MY_INFORMATION,
  GET_MY_DEPARTMENT_INFO,
  GET_MY_ORGANISATION_INFO,
};
